File: pkg-r/R/TableSet.R
# The tables a chat can query, plus the system prompt and query executor
# built from them. Never mutated after construction: QueryChat holds one
# built from $add_table() calls, and $server(data_source = ) derives a second
# one per session so a running session never observes later changes.
#
# Not exported.
TableSet <- R6::R6Class(
  "TableSet",
  private = list(
    .data_sources = NULL,
    .system_prompt = NULL,
    .data_description = NULL,
    .executor = NULL
  ),
  public = list(
    initialize = function(
      data_sources,
      system_prompt,
      data_description = NULL
    ) {
      if (length(data_sources) == 0) {
        cli::cli_abort("{.cls TableSet} requires at least one data source.")
      }
      nms <- names(data_sources)
      if (is.null(nms) || any(!nzchar(nms))) {
        cli::cli_abort("{.arg data_sources} must be a named list.")
      }
      private$.data_sources <- data_sources
      private$.system_prompt <- system_prompt
      private$.data_description <- data_description
    },

    executor = function() {
      if (is.null(private$.executor)) {
        private$.executor <- build_query_executor(private$.data_sources)
      }
      private$.executor
    },

    executor_built = function() {
      !is.null(private$.executor)
    },

    table_names = function() {
      names(private$.data_sources)
    },

    # Closes the executor if it was ever built. Never touches data sources.
    cleanup_executor = function() {
      if (!is.null(private$.executor)) {
        tryCatch(
          private$.executor$cleanup(),
          finally = {
            private$.executor <- NULL
          }
        )
      }
      invisible(NULL)
    }
  ),
  active = list(
    data_sources = function(value) {
      if (!missing(value)) {
        cli::cli_abort("{.field data_sources} is read-only.")
      }
      private$.data_sources
    },
    system_prompt = function(value) {
      if (!missing(value)) {
        cli::cli_abort("{.field system_prompt} is read-only.")
      }
      private$.system_prompt
    },
    data_description = function(value) {
      if (!missing(value)) {
        cli::cli_abort("{.field data_description} is read-only.")
      }
      private$.data_description
    }
  )
)

File: pkg-py/src/querychat/_table_set.py
"""Immutable bundle of the tables a chat can query."""

from __future__ import annotations

from functools import cached_property
from types import MappingProxyType
from typing import TYPE_CHECKING, Generic

from narwhals.stable.v1.typing import IntoFrameT

from ._query_executor import QueryExecutor, build_query_executor

if TYPE_CHECKING:
    from collections.abc import Mapping

    from ._datasource import DataSource
    from ._system_prompt import QueryChatSystemPrompt


class TableSet(Generic[IntoFrameT]):
    """
    The tables a chat can query, plus the prompt and executor built from them.

    A ``TableSet`` is never mutated after construction. ``QueryChatBase``
    holds one built from ``add_table()`` calls; ``QueryChat.server()`` derives
    a second one when a session registers its own table, so a running session
    never observes changes made after it started.
    """

    def __init__(
        self,
        data_sources: Mapping[str, DataSource[IntoFrameT]],
        system_prompt: QueryChatSystemPrompt,
    ) -> None:
        if not data_sources:
            raise ValueError("TableSet requires at least one data source")
        self._data_sources: Mapping[str, DataSource[IntoFrameT]] = MappingProxyType(
            dict(data_sources)
        )
        self._system_prompt = system_prompt

    @property
    def data_sources(self) -> Mapping[str, DataSource[IntoFrameT]]:
        return self._data_sources

    @property
    def system_prompt(self) -> QueryChatSystemPrompt:
        return self._system_prompt

    @cached_property
    def executor(self) -> QueryExecutor:
        return build_query_executor(self.data_sources)

    @property
    def executor_built(self) -> bool:
        return "executor" in self.__dict__

    @property
    def table_names(self) -> list[str]:
        return list(self.data_sources)

    def cleanup_executor(self) -> None:
        """
        Close the executor if it was ever built. Never touches data sources.

        The cached executor is reset (matching R's ``TableSet``), so a later
        access rebuilds it.
        """
        if self.executor_built:
            try:
                self.executor.cleanup()
            finally:
                del self.__dict__["executor"]
